src/struct_mv/struct_matmult_core.c

```diff
@@ -18,56 +18,467 @@
+ *    building on the previous ones.
  *
- * F/C means fine/coarse data space.  For example, CFF means the data spaces for
- * the three terms are respectively coarse, fine, and fine.
+ *    The macro concatenation (##) is used to dispatch to the appropriate
+ *    implementation based on the requested depth. This technique allows us
+ *    to write generic code that selects the correct number of unrolled loops
+ *    at compile time.
+ *
+ * 3. Computation core macros:
+ *    HYPRE_SMMCORE_* - Define the actual multiplication operations
+ *    - Different variants for different data space combinations (FFF, CFF, etc.)
+ *    - Each computes a product of coefficient and data values
+ *
+ *    HYPRE_SMMCORE_*_SUM_UP_TO_n - Create unrolled sums of multiple products at compile time
+ *    - Direct entry points for generating compile-time sums of core computations
+ *    - The n suffix determines how many terms to include (from 0 to n-1)
+ *    - Used inside computation loops after loading variables with HYPRE_SMMLOAD_*_UP_TO
+ *
+ * Usage pattern in computation:
+ * -----------------------------
+ * 1. Declare variables using HYPRE_DECLARE_ALL_*VARS
+ * 2. For each stencil entry:
+ *    a. Load data using HYPRE_SMMLOAD_*VARS_UP_TO(k, n) where n is depth
+ *    b. Compute products using HYPRE_SMMCORE_* macros via BoxLoops
+ *    c. Accumulate results into the output pointer (mptr)
  *--------------------------------------------------------------------------*/
 
+#define HYPRE_DECLARE_2VARS(k) \
+   HYPRE_Complex *tptrs_##k##_0, cprod_##k
+
+#define HYPRE_DECLARE_3VARS(k) \
+   HYPRE_Complex *tptrs_##k##_0, *tptrs_##k##_1, cprod_##k
+
+#define HYPRE_DECLARE_4VARS(k) \
+   HYPRE_Complex *tptrs_##k##_0, *tptrs_##k##_1, *tptrs_##k##_2, cprod_##k
+
+#define HYPRE_DECLARE_ALL_2VARS \
+   HYPRE_DECLARE_2VARS(0);      \
+   HYPRE_DECLARE_2VARS(1);      \
+   HYPRE_DECLARE_2VARS(2);      \
+   HYPRE_DECLARE_2VARS(3);      \
+   HYPRE_DECLARE_2VARS(4);      \
+   HYPRE_DECLARE_2VARS(5);      \
+   HYPRE_DECLARE_2VARS(6)
+
+#define HYPRE_DECLARE_ALL_3VARS \
+   HYPRE_DECLARE_3VARS(0);      \
+   HYPRE_DECLARE_3VARS(1);      \
+   HYPRE_DECLARE_3VARS(2);      \
+   HYPRE_DECLARE_3VARS(3);      \
+   HYPRE_DECLARE_3VARS(4);      \
+   HYPRE_DECLARE_3VARS(5);      \
+   HYPRE_DECLARE_3VARS(6)
+
+#define HYPRE_DECLARE_ALL_4VARS \
+   HYPRE_DECLARE_4VARS(0);      \
+   HYPRE_DECLARE_4VARS(1);      \
+   HYPRE_DECLARE_4VARS(2);      \
+   HYPRE_DECLARE_4VARS(3);      \
+   HYPRE_DECLARE_4VARS(4);      \
+   HYPRE_DECLARE_4VARS(5);      \
+   HYPRE_DECLARE_4VARS(6)
+
+#define HYPRE_SMMLOAD_2VARS(k, m)     \
+   cprod_##m     = cprod[(k + m)];    \
+   tptrs_##m##_0 = tptrs[(k + m)][0]
+
+#define HYPRE_SMMLOAD_3VARS(k, m)     \
+   cprod_##m     = cprod[(k + m)];    \
+   tptrs_##m##_0 = tptrs[(k + m)][0]; \
+   tptrs_##m##_1 = tptrs[(k + m)][1]
+
+#define HYPRE_SMMLOAD_4VARS(k, m)     \
+   cprod_##m     = cprod[(k + m)];    \
```

Comment:
Why do we need to lose the k-array structure here?  Any idea if this affects the CPU runs?

Reply:
This is because we need to pass device pointers to the BoxLoops in the case of GPU runs, and the double pointers like tptrs can be an issue because if any of the inner pointers reside on the host (or were set up on the host), they become invalid when dereferenced on the device. So, here the idea is to reduce the level of indirection and create variables to point directly to device addresses.

I don't expect this change to affect the performance on CPU runs, but I will double check with experiments.

Reply:
Thanks for the explanation, and just do a quick performance check for a single rank run.  No need to do anything extensive because it's probably all okay.